stake-pool: Stake pool deserializer

lieuzhenghong opened the pull request:
This PR allows successful deserialisation of `StakePool` and `ValidatorList` account data. I wrote Borsh.js schema and Typescript bindings in `schema.ts`, and some helper functions in `index.ts` to try out the bindings. 

For consistency, I've used the same Prettier settings as the one used in `token/`.

**NOTE:** We want to get `borsh-js` to release a new version ASAP (the one that includes the u128/u256/u512 methods by @joncinque), because currently the import in `package.json` relies on having borsh on your local machine, which is really not ideal:

```package.json
"borsh": "file:../../../solana-repos/borsh-js",
```

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,104 @@
+        },
+      };
+
+      if (a.account.data.length === STAKE_POOL_ACCT_LENGTH) {
```
We shouldn't have to do this, it defeats the whole purpose of a deserializing library like borsh!  I see two possible situations:
* we know the type being deserialized, so we just use it's decode right away
* we really don't know, in which case `AccountType` can actually have a value of `StakePool` / `ValidatorList`. You would just need to omit the account type enum at the start.  We shouldn't really run into that situation though.  If we need to do `getProgramAccounts`, we can use a filter on the first byte, either it equals 1 (StakePool), or 2 (ValidatorList)

lieuzhenghong replied:
Fair enough --- love the idea to filter on the first byte. I've done that now.

joncinque commented on stake-pool/js/src/schema.js:
```diff
@@ -0,0 +1,126 @@
+import borsh from 'borsh';
```
Same here, why js and ts?

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,167 @@
+  manager: PublicKey;
+  staker: PublicKey;
+  depositAuthority: PublicKey;
+  withdrawBumpSeed: number; // what is this? u8 in Rust
```
number's good here

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,167 @@
+  reserveStake: PublicKey;
+  poolMint: PublicKey;
+  managerFeeAccount: PublicKey;
+  totalStakeLamports: number;
+  poolTokenSupply: number;
+  lastUpdateEpoch: number;
```
These should be BN, since node doesn't support 64-bit ints

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,167 @@
+export class ValidatorStakeInfo extends Assignable {
+  status: StakeStatus;
+  voteAccountAddress: PublicKey;
+  stakeLamports: number;
+  lastUpdateEpoch: number;
```
Same with these

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,167 @@
+}
+
+export class Fee extends Assignable {
+  denominator: number;
+  numerator: number;
```
These should be ok as number

lieuzhenghong replied:
why is this OK as number?

joncinque replied:
Although they're both `u64`s underneath, people *most likely* wouldn't use something so big.  To be totally safe, they should be `BN`s too, but it probably isn't important

joncinque reviewed (commented):
The structs themselves look good!  Mainly some comments on the client side

lieuzhenghong commented on stake-pool/js/src/index.js:
```diff
@@ -0,0 +1,64 @@
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
```
Yes, the `.js` files aren't needed. These are built by `npx tsc` -- will remove them.

lieuzhenghong commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,104 @@
+   * Each stake pool has a stake pool account and a validator list account.
+   * (Optionally) a stake pool can also have a name and ticker.
+   */
+  name: string;
+  ticker: string;
```
These aren't stake pools per se, these are "combined" stake pools that have both stake pool and validator list accounts. Dom wants each stake pool to have a human-readable name and ticker. I can remove them if you'd like?

joncinque replied:
Hm, seems like this part should be provided by another service.  Where are the human-readable name / ticker supposed to come from?  If we have a name anywhere, I thought it would be on the mint for the stake pool token using one of the token metadatas.

lieuzhenghong replied:
I think the idea is for stake pool managers to give us their desired stake pool name and ticker. Using the token metadata is a good idea.

lieuzhenghong commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,104 @@
+  return accountType.decode(serializedStakePool);
+}
+
+async function getStakePoolAccounts(
```
Yep, I agree. I have written two such functions: `getStakePoolAccount` and `getValidatorListAccount`.

lieuzhenghong commented:
I've incorporated all of the requested changes from Jon:

- Removed extraneous `.js` files
- Changed schema 
- Stopped using `a.account.data.length` and instead filtered on the first byte to distinguish between `StakePool` and `ValidatorList`
- Wrote individual functions to get a single `StakePool` or `ValidatorList` account

And a few of my own:

- Added two helper interfaces, `decodedStakePool` and `decodedValidatorList`.
- Wrote helper functions to pretty print public keys and decoded accounts
- Cleaned up the code

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,168 @@
+    const stakePoolAccounts = response.map(a => {
+      let decodedData;
+
+      if (a.account.data.readUInt8() === 1) {
```
For a next phase (v2?), we can do this even better :smile: we should probably have all of our state structs given as:
```rs
pub struct State {
    Uninitialized,
    StakePool(StakePool),
    ValidatorList(ValidatorList),
}
```
And then we can just call deserialize and we're good

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,166 @@
+  managerFeeAccount: PublicKey;
+  totalStakeLamports: BN;
+  poolTokenSupply: BN;
+  lastUpdateEpoch: BN;
```
This is missing `Fee`

lieuzhenghong replied:
Added

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,166 @@
+  numerator: number;
+}
+
+export class AccountType extends Enum {}
```
This is currently not usable, correct? Can you add the enum variants?

lieuzhenghong replied:
Resolved with TypeScript enums, as discussed on Discord: https://discord.com/channels/428295358100013066/766355323580579901/839204220124659783

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,166 @@
+  stakeLamports: BN;
+  lastUpdateEpoch: BN;
+}
+export class StakeStatus extends Enum {}
```
Same here, this needs the enum variants

joncinque commented:
Looking close, just a few more comments! Then we'll need to wait for the borsh stuff to be integrated, let me ping on that side

lieuzhenghong commented:
1. Fix `borshjs` import (**blocking**: if the `near` folks don't release a new version, we may have to publish our own package)
2. Write unit and integration tests 
2. Fill in fields in `package.json`

lieuzhenghong commented:
@joncinque , it looks like near still hasn't released a new version -- what do we have to do?

joncinque commented:
They actually did release their new version!  And it was integrated with the newest web3.js in https://github.com/solana-labs/solana/pull/17239, so you can go ahead and use all of the newest stuff

lieuzhenghong commented:
I've upgraded the `borsh` import, wrote tests to verify correctness of `schema.decode`, and filled in fields in `package.json`. I think I'm ready to merge.

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,172 @@
+  validatorList: decodedValidatorList;
+}
+
+export interface decodedStakePool {
```
I think the type names can be reversed:
```suggestion
export interface StakePoolAccount {
```

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,172 @@
+  account: solanaWeb3.AccountInfo<schema.StakePoolAccount>;
+}
+
+export interface decodedValidatorList {
```
```suggestion
export interface ValidatorListAccount {
```

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,172 @@
+ * Also contains several useful helper functions
+ */
+
+export class StakePool {
```
This can be `StakePoolAccounts`?

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,172 @@
+  console.log('Owner PubKey:', account.account.owner.toString());
+}
+
+export default function testOnDevnet() {
```
Should this be in the test dir?

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+import BN from 'bn.js';
+
+// Class wrapping a plain object
+export abstract class Assignable {
```
You can remove this and use `Struct` https://github.com/solana-labs/solana/blob/ec1a307a7c2618b9bdea7b82a39a36ba194ae0d0/web3.js/src/util/borsh-schema.ts#L5

lieuzhenghong replied:
I need to use `deserializeUnchecked` in the `decode` method or validatorlist won't decode properly. Is it possible to change this in `borsh-schema.ts`?

joncinque replied:
Ah right very good point.  Could you add another function `decodeUnchecked` in `borsh-schema.ts` which calls `deserializeUnchecked` under the hood?  Tag me on the PR and I'll review it as quickly as possible.

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+
+// Class representing a Rust-compatible enum, since enums are only strings or
+// numbers in pure JS
+export abstract class Enum extends Assignable {
```
You can remove this and use `Enum` from web3.js https://github.com/solana-labs/solana/blob/ec1a307a7c2618b9bdea7b82a39a36ba194ae0d0/web3.js/src/util/borsh-schema.ts#L21

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+  ValidatorList = 'ValidatorList',
+}
+
+export class StakePoolAccount extends Assignable {
```
And this can be renamed to `StakePool`

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+  fee: Fee;
+}
+
+export class ValidatorListAccount extends Assignable {
```
This renamed to `ValidatorList`

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+  ReadyForRemoval = 'ReadyForRemoval',
+}
+
+export class PublicKey extends Assignable {
+  value: BN;
+}
```
This can be removed in favor of the web3.js PublicKey

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,181 @@
+  value: BN;
+}
+
+export const SCHEMA: Schema = constructStakePoolSchema();
```
You can add to `SOLANA_SCHEMA` defined at https://github.com/solana-labs/solana/blob/ec1a307a7c2618b9bdea7b82a39a36ba194ae0d0/web3.js/src/util/borsh-schema.ts#L34

joncinque commented on stake-pool/js/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+        "outDir": "dist"
+    },
+    "include": [
+        "src/*.ts"
```
```suggestion
        "src/**/*.ts"
```

joncinque reviewed (commented):
Looking really close, just a few more bits to take care of!

lieuzhenghong commented:
- Used new version of borsh.js in PR https://github.com/solana-labs/solana/pull/17620 (thanks @joncinque)
- Refactored type names
- Add new `testOnDevnet` test
- Upgrade web3js version

```
  schema.decode
    StakePoolAccount
      ✓ should successfully decode StakePoolAccount account data
    ValidatorListAccount
      ✓ should successfully decode ValidatorListAccount account data
      ✓ should successfully decode ValidatorListAccount with nonempty ValidatorInfo

  index.ts/PrettyPrintPubkey
    ✓ should successfully pretty print a pubkey

  testOnDevnet
    ✓ should successfully decode all validators from devnet (476ms)


  5 passing (862ms)
  ```

joncinque commented on stake-pool/js/package.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "stake-pool-js",
+  "version": "0.0.1",
```
We can probably start on version 0.1.0
```suggestion
  "version": "0.1.0",
```

joncinque commented on stake-pool/js/package.json:
```diff
@@ -0,0 +1,28 @@
+  "scripts": {
+    "lint": "npx prettier src/*.ts -w",
+    "compile": "tsc",
+    "test": "./node_modules/mocha/bin/mocha -p ./dist"
```
We should be able to just call `mocha`, but this probably requires a `build` step beforehand
```suggestion
    "test": "mocha -p ./dist"
```

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,167 @@
+  connection: solanaWeb3.Connection,
+  stakePoolPubKey: solanaWeb3.PublicKey,
+): Promise<StakePoolAccount> {
+  /**
+   * Retrieves and deserializes a StakePool account using a web3js connection and the stake pool address.
+   * @param connection: An active web3js connection.
+   * @param stakePoolPubKey: The public key (address) of the stake pool account.
+   */
```
I always thought these comments had to be before the code to apply correctly, e.g. 
```
/**
 * my comment
 */
export async function myFunction()
```
Can it also be in this position?

joncinque commented on stake-pool/js/src/index.ts:
```diff
@@ -0,0 +1,167 @@
+   * @param connection: An active web3js connection.
+   * @param stakePoolPubKey: The public key (address) of the stake pool account.
+   */
+  try {
```
we probably don't need to wrap the whole function in a try-catch block

lieuzhenghong replied:
done in https://github.com/solana-labs/solana-program-library/pull/1656/commits/87522abe8ad87806d24fd071171b2548c4483e25

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,147 @@
+  ReadyForRemoval = 'ReadyForRemoval',
+}
+
+export class PublicKey extends Struct {
```
PublicKey is borsh-compatible, so feel free to remove this and use the web3 version everywhere!

lieuzhenghong replied:
done in https://github.com/solana-labs/solana-program-library/pull/1656/commits/1970f0d7f88969c718bca6ec07829191eec9de6a

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,147 @@
+
+export const SCHEMA: Schema = constructStakePoolSchema(SOLANA_SCHEMA);
+
+export function constructStakePoolSchema(SCHEMA: Schema): Schema {
```
nit, all caps variable names are typically globals
```suggestion
export function constructStakePoolSchema(schema: Schema): Schema {
```

joncinque commented on stake-pool/js/src/schema.ts:
```diff
@@ -0,0 +1,147 @@
+  value: BN;
+}
+
+export const SCHEMA: Schema = constructStakePoolSchema(SOLANA_SCHEMA);
```
nit: This is a bit of a strange function, since it's modifying `SOLANA_SCHEMA` in place and then also returning it.  I would say the function should either clone the map, then modify and return, OR just modify it in place.

Take it or leave it, but to me the best bet would be to rename the function `addStakePoolSchema` and modify SOLANA_SCHEMA in place.

lieuzhenghong replied:
Makes sense: have made the changes in https://github.com/solana-labs/solana-program-library/pull/1656/commits/3e8073bf1f2a560d5a746175cef3c4c1d21510a9

joncinque commented on stake-pool/js/src/test.ts:
```diff
@@ -0,0 +1,194 @@
+    it('should successfully decode StakePoolAccount account data', () => {
+      const decodedData = schema.StakePool.decode(
+        new Buffer(
+          '014f962536a1b5b3a1f1036599ca7c164d6ad0677fed896f12ecdd09ada2cd23dc4f962536a1b5b3a1f1036599ca7c164d6ad0677fed896f12ecdd09ada2cd23dcbb09705cd25f4ad3b9e2c78e3f62ba3e290f798150499edc4f851a1a45e71159ff1086831390e666d64d18cac8f0c86eb91d799eda647319634280b89548ba0371ee618745ef3ebb93164e66ac3560e934b04e49b89ad38baa3b104284abb1a574003e566772f55676cf243c647fab463afa138cb156d2080062d2957b10bf2287 ...
```
this test may be slightly brittle... I've since added some fields, which would break these tests, and modifying them is pretty annoying.  Maybe you can test the round-trip logic of creating the type, serializing, and deserializing?

lieuzhenghong replied:
Not sure what you mean here: wouldn't adding fields break the entire deserialiser, not just the tests?

lieuzhenghong replied:
solved in https://github.com/solana-labs/solana-program-library/pull/1656/commits/e122e32c348ef0ec449b11e0715e32a17092e9bc

joncinque commented on stake-pool/js/src/test.ts:
```diff
@@ -0,0 +1,194 @@
+    it('should successfully decode ValidatorListAccount account data', () => {
+      const decodedData = schema.ValidatorList.decodeUnchecked(
+        new Buffer(
+          '020a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ...
```
We should include some data in this to make it test the full deserialization behavior

lieuzhenghong replied:
Done in https://github.com/solana-labs/solana-program-library/pull/1656/commits/01f557a157dcf11837dc097f21df1ce4bd42c23c

joncinque commented on stake-pool/js/src/test.ts:
```diff
@@ -0,0 +1,194 @@
+  });
+});
+
+describe('Integration test', () => {
```
We can keep this for now, but when this package is integrated in the CI, we should move these tests into a different place as a separate suite of tests.  This way, we'll be sure to never break backwards compatibility :smile:

lieuzhenghong replied:
done in https://github.com/solana-labs/solana-program-library/pull/1656/commits/02df5acc687bf04bea38c4033cf81513a20c7dcd

joncinque commented on stake-pool/js/webpack.config.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
```
Is this used?

lieuzhenghong replied:
removed in https://github.com/solana-labs/solana-program-library/pull/1656/commits/0922acc76c05eaea559901e141034d2609db0a56

joncinque reviewed (commented):
Looks really close! Just a few small things to work out, then this should be good to integrate.  Once it's in, then we can add support for the instructions and maybe some end-to-end tests setting up a pool and all that!

joncinque commented:
All right, let's get this in!  We can do more tweaks for the updated fields and integrate CI in later work.  Thanks for your contribution!
